File: AEDS/BASES/Arvore/arvoreAN.java
public class arvoreAN {
    public class No {
        private int elemento;
        private boolean cor;
        private No esq, dir;
        public No (int x) {
            this.elemento = x;
            this.cor = false;
            this.esq = this.dir = null;
        }
        public No (int x, No esq, No dir) {
            this.elemento = x;
            this.cor = false;
            this.esq = esq;
            this.dir = dir;
        }
    }

    public class Arvore {
        private No raiz;
        public Arvore () {
            this.raiz = null;
        }

        private No inserirRef (int x, No i) {
            if (i == null) i = new No(x);
            else if (x < i.elemento) i.esq = inserirRef(x, i.esq);
            else if (x > i.elemento) i.dir = inserirRef(x, i.dir);
            else throw new RuntimeException ("Erro!");
            return balancear(i);
        }
        public void inserir (int x) {
            raiz = inserirRef(x, raiz);
        }

        private No balancear (No i) {
            if (i != null) {
                int fator = i.getNivel(i.dir) - i.getNivel(i.esq);
                if (Math.abs(fator) <= 1) i.setNivel();
                else if (fator == 2) {
                    int fatorFilhoDir = i.dir.getNivel(i.dir.dir) - i.dir.getNivel(i.dir.esq);
                    if (fatorFilhoDir == -1) i.dir = rotacionarDir(i.dir);
                    i = rotacionarEsq(i);
                }
                else if (fator == -2) {
                    int fatorFilhoEsq = i.esq.getNivel(i.esq.dir) - i.esq.getNivel(i.esq.esq);
                    if (fatorFilhoEsq == 1) i.esq = rotacionarEsq(i.esq);
                    i = rotacionarDir(i);
                } else throw new RuntimeException ("Erro!");
            }
            return i;
        }

        private No rotacionarEsq (No i) {
            No noDir = i.dir;
            No noDirEsq = noDir.esq;
            noDir.esq = i;
            i.dir = noDirEsq;
            i.setNivel();
            noDir.setNivel();
            return noDir;
        }
        private No rotacionarDir (No i) {
            No noEsq = i.esq;
            No noEsqDir = noEsq.dir;
            noEsq.dir = i;
            i.esq = noEsqDir;
            i.setNivel();
            noEsq.setNivel();
            return noEsq;
        }

        private void inserirPai (int x, No i, No pai) {
            if (i == null) {
                if (x < pai.elemento) pai.esq = new No (x);
                else pai.dir = new No (x);
            }
            else if (x < i.elemento) inserirPai(x, i.esq, i);
            else if (x > i.elemento) inserirPai(x, i.dir, i);
            else throw new RuntimeException ("Erro!");
        }
        public void inserirPai (int x) {
            if (raiz == null) raiz = new No (x);
            else if (x < raiz.elemento) inserirPai(x, raiz.esq, raiz);
            else if (x > raiz.elemento) inserirPai(x, raiz.dir, raiz);
            else throw new RuntimeException ("Erro!");
        }

        private boolean pesquisar (int x, No i) {
            if (i == null) return false;
            else if (x == i.elemento) return true;
            else if (x < i.elemento) return pesquisar(x, i.esq);
            else return pesquisar(x, i.dir);
        }
        public boolean pesquisar (int x) {
            return pesquisar(x, raiz);
        }

        public void caminharCentral (No i) {
            if (i != null) {
                caminharCentral(i.esq);
                System.out.print(i.elemento + " ");
                caminharCentral(i.dir);
            }
        }
        public void caminharPre (No i) {
            if (i != null) {
                System.out.print(i.elemento + " ");
                caminharPre(i.esq);
                caminharPre(i.dir);
            }
        }
        public void caminharPos (No i) {
            if (i != null) {
                caminharPos(i.esq);
                caminharPos(i.dir);
                System.out.print(i.elemento + " ");
            }
        }

        private No remover (int x, No i) {
            if (i == null) throw new RuntimeException ("Erro!");
            else if (x < i.elemento) i.esq = remover(x, i.esq);
            else if (x > i.elemento) i.dir = remover(x, i.dir);
            else if (i.dir == null) i = i.esq;
            else if (i.esq == null) i = i.dir;
            else i.esq = maiorEsq(i, i.esq);
            return balancear(i);
        }
        public void remover (int x) {
            raiz = remover(x, raiz);
        }

        private No maiorEsq (No i, No j) {
            if (j.dir == null) {
                i.elemento = j.elemento;
                j = j.esq;
            }
            else j.dir = maiorEsq(i, j.dir);
            return j;
        }
        private No menorDir (No i, No j) {
            if (j.esq == null) {
                i.elemento = j.elemento;
                j =j.dir;
            }
            else j.esq = menorDir(i, j.esq);
            return j;
        }

        public int getAlturaRec (No i) {
            if (i == null) return 0;
            else {
                int altEsq = getAlturaRec(i.esq);
                int altDir = getAlturaRec(i.dir);
                return (altEsq > altDir ? altEsq : altDir) + 1;
            }
        }

        public int getMaior () {
            if (raiz == null) throw new RuntimeException ("Erro!");
            No i;
            for (i = raiz; i.dir != null; i = i.dir);
            return i.elemento;
        }
        public int getMenor () {
            if (raiz == null) throw new RuntimeException ("Erro!");
            No i;
            for (i = raiz; i.esq != null; i = i.esq);
            return i.elemento;
        }

        public int somarRec (No i) {
            if (i == null) return 0;
            else return i.elemento + somarRec(i.esq) + somarRec(i.dir);
        }

        private boolean igual (No a, No b) {
            if (a == null && b == null) return true;
            else if (a == null || b == null) return false;
            else if (a.elemento != b.elemento) return false;
            else return igual(a.esq, b.esq) && igual(a.dir, b.dir);
        }
        public boolean igual (Arvore b) {
            return igual(this.raiz, b.raiz);
        }
        private boolean espelho (No a, No b) {
            if (a == null && b == null) return true;
            else if (a == null || b == null) return false;
            else if (a.elemento != b.elemento) return false;
            else return espelho(a.esq, b.dir) && espelho(a.dir, b.esq);
        }
        public boolean espelho (Arvore b) {
            return espelho(this.raiz, b.raiz);
        }
    }
}

File: AEDS/BASES/Arvore/arvoreAN.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct NoAN {
    int elemento;
    bool cor;
    struct NoAN *esq, *dir;
} NoAN;

NoAN *newNo (int x) {
    NoAN *novo = (NoAN*)malloc(sizeof(NoAN));
    novo->elemento = x;
    novo->cor = false;
    novo->esq = novo->dir = NULL;
    return novo;
}

NoAN *raiz;

void ArvoreAN () {
    raiz = NULL;
}

NoAN* rotacionarDir (NoAN *i) {
    NoAN *noEsq = i->esq;
    NoAN *noEsqDir = noEsq->dir;
    noEsq->dir = i;
    i->esq = noEsqDir;
    return noEsq;
}
NoAN* rotacionarEsq (NoAN *i) {
    NoAN *noDir = i->dir;
    NoAN *noDirEsq = noDir->esq;
    noDir->esq = i;
    i->dir = noDirEsq;
    return noDir;
}
NoAN* rotacionarDirEsq (NoAN *i) {
    i->dir = rotacionarDir(i->dir);
    return rotacionarEsq(i);
}
NoAN* rotacionarEsqDir (NoAN *i) {
    i->esq = rotacionarEsq(i->esq);
    return rotacionarDir(i);
}

void balancear (NoAN *bis, NoAN *avo, NoAN *pai, NoAN *i) {
    if (pai->cor == true) {
        if (pai->elemento > avo->elemento) {
            if (i->elemento > pai->elemento) avo = rotacionarEsq(avo);
            else avo = rotacionarDirEsq(avo);
        } else {
            if (i->elemento < pai->elemento) avo = rotacionarDir(avo);
            else avo = rotacionarEsqDir(avo);
        }
        if (bis == NULL) raiz = avo;
        else if (avo->elemento < bis->elemento) bis->esq = avo;
        else bis->dir = avo;
        avo->cor = false;
        avo->esq->cor = avo->dir->cor = true;
    }
}

void inserirRec (int x, NoAN *bis, NoAN *avo, NoAN *pai, NoAN *i) {
    if (i == NULL) {
        if (x < pai->elemento) i = pai->esq = newNo(x);
        else i = pai->dir = newNo(x);
        if (pai->cor == true) balancear(bis, avo, pai, i);
    } else {
        if (i->esq != NULL && i->dir != NULL && i->esq->cor == true && i->dir->cor == true) {
            i->cor = true;
            i->esq->cor = i->dir->cor = false;
            if (i == raiz) i->cor = false;
            else if (pai->cor == true) balancear(bis, avo, pai, i);
        }
        if (x < i->elemento) inserirRec(x, pai, avo, i, i->esq);
        else if (x > i->elemento) inserirRec(x, pai, avo, i, i->dir);
        else exit(1);
    }
}
void inserir (int x) {
    if (raiz == NULL) raiz = newNo(x);
    else if (raiz->esq == NULL && raiz->dir == NULL) {
        if (x < raiz->elemento) raiz->esq = newNo(x);
        else if (x > raiz->elemento) raiz->dir = newNo(x);
    } else if (raiz->esq == NULL) {
        if (x < raiz->elemento) raiz->esq = newNo(x);
        else if (x < raiz->dir->elemento) {
            raiz->esq = newNo(raiz->elemento);
            raiz->elemento = x;
        } else {
            raiz->esq = newNo(raiz->elemento);
            raiz->elemento = raiz->dir->elemento;
            raiz->dir->elemento = x;
        }
        raiz->esq->cor = raiz->dir->cor = false;
    } else if (raiz->dir == NULL) {
        if (x > raiz->elemento) raiz->dir = newNo(x);
        else if (x > raiz->esq->elemento) {
            raiz->dir = newNo(raiz->elemento);
            raiz->elemento = x;
        } else {
            raiz->dir = newNo(raiz->elemento);
            raiz->elemento = raiz->esq->elemento;
            raiz->esq->elemento = x;
        }
        raiz->esq->cor = raiz->dir->cor = false;
    } else inserirRec(x, NULL, NULL, NULL, raiz);
    raiz->cor = false;
}

bool pesquisarNo (NoAN *i, int x) {
    if (i == NULL) return false;
    else if (x == i->elemento) return true;
    else if (x < i->elemento) return pesquisarNo(i->esq, x);
    else return pesquisarNo(i->dir, x);
}
bool pesquisar (int x) {
    return pesquisarNo(raiz, x);
}

NoAN* removerNo (int x, NoAN *i) {
    if (i == NULL) exit(1);
    else if (x < i->elemento) i->esq = removerNo(x, i->esq);
    else if (x > i->elemento) i->dir = removerNo(x, i->dir);
    else if (i->dir == NULL) i = i->esq;
    else if (i->esq == NULL) i = i->dir;
    else i->esq = maiorEsq(i, i->esq);
    return i;
}
void remover (int x) {
    raiz = removerNo(x, raiz);
}

NoAN* maiorEsq (NoAN *i, NoAN *j) {
    if (j->dir == NULL) {
        i->elemento = j->elemento;
        j = j->esq;
    }
    else j->dir = maiorEsq(i, j->dir);
    return j;
}
NoAN* menorDir (NoAN *i, NoAN *j) {
    if (j->esq == NULL) {
        i->elemento = j->elemento;
        j = j->dir;
    }
    else j->esq = menorDir(i, j->esq);
    return j;
}

void caminharCentral (NoAN *i) {
    if (i != NULL) {
        caminharCentral(i->esq);
        printf("%d %s ", i->elemento, (i->cor ? "(P)" : "(B)"));
        caminharCentral(i->dir);
    }
}
void caminharPre (NoAN *i) {
    if (i != NULL) {
        printf("%d %s ", i->elemento, (i->cor ? "(P)" : "(B)"));
        caminharPre(i->esq);
        caminharPre(i->dir);
    }
}
void caminharPos (NoAN *i) {
    if (i != NULL) {
        caminharPos(i->esq);
        caminharPos(i->dir);
        printf("%d %s ", i->elemento, (i->cor ? "(P)" : "(B)"));
    }
}

int getAlturaRec (NoAN *i) {
    if (i == NULL) return 0;
    else {
        int altEsq = getAlturaRec(i->esq);
        int altDir = getAlturaRec(i->dir);
        return (altEsq > altDir ) ? altEsq + 1: altDir + 1;
    }
}

int getMaior () {
    if (raiz == NULL) exit(1);
    NoAN *i;
    for (i = raiz; i->dir != NULL; i = i->dir);
    return i->elemento;
}
int getMenor () {
    if (raiz == NULL) exit(1);
    NoAN *i;
    for (i = raiz; i->esq != NULL; i = i->esq);
    return i->elemento;
}

int somarRec (NoAN *i) {
    if (i == NULL) return 0;
    return i->elemento + somarRec(i->esq) + somarRec(i->dir);
}

bool igual (NoAN *a, NoAN *b) {
    if (a == NULL && b == NULL) return true;
    else if (a == NULL || b == NULL) return false;
    else if (a->elemento != b->elemento) return false;
    else return igual(a->esq, b->esq) && igual(a->dir, b->dir);
}
bool igualArvore (NoAN *outraRaiz) {
    return igual(raiz, outraRaiz);
}
bool espelho (NoAN *a, NoAN *b) {
    if (a == NULL && b == NULL) return true;
    else if (a == NULL || b == NULL) return false;
    else if (a->elemento != b->elemento) return false;
    else return espelho(a->esq, b->dir) && espelho(a->dir, b->esq);
}
bool espelhoArvore (NoAN *outraRaiz) {
    return espelho(raiz, outraRaiz);
}
